File: Python3/0078_subsets.py
# https://leetcode.cn/problems/subsets/

class Solution:
    def subsets(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: list[list[int]]
        """

        self.res = []
        n = len(nums)
        self.backtrack(nums, n, [], 0)
        return self.res

    def backtrack(self, nums, n, tmp, idx):
        self.res.append(tmp)

        for i in range(idx, n):
            self.backtrack(nums, n, tmp+[i+1], i+1)

File: Cpp/0078_subsets.cpp
// https://leetcode.cn/problems/subsets/

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> tmp;
        int n = nums.size();
        backtrack(nums, tmp, n, 0);
        return res;
    }

private:
    void backtrack(vector<int> nums, vector<int>& tmp, int n, int idx) {
        res.push_back(tmp);

        for (int i = idx; i < n; ++i) {
            tmp.push_back(nums[i]);
            backtrack(nums, tmp, n, i+1);
            tmp.pop_back();
        }
    }

    vector<vector<int>> res;
};
